kotlinx-coroutines-core/common/src/CoroutineDispatcher.kt

```diff
@@ -27,46 +27,30 @@ import kotlin.coroutines.*
-     * line, while the UI update happens asynchronously in the UI thread. However, if you invoke
-     * it in the UI thread itself, it will update the UI _synchronously_ if your `isDispatchNeeded` is
-     * overridden with a thread check. Checking if we are already in the UI thread seems more
-     * efficient (and it might indeed save a few CPU cycles), but this subtle and context-sensitive
-     * difference in behavior makes the resulting async code harder to debug.
+     * If this method returns `false`, the coroutine is resumed immediately in the current thread,
+     * potentially forming an event-loop to prevent stack overflows.
+     * The event loop is an advanced topic and its implications can be found in [Dispatchers.Unconfined] documentation.
      *
-     * Basically, the choice here is between the "JS-style" asynchronous approach (async actions
-     * are always postponed to be executed later in the event dispatch thread) and "C#-style" approach
-     * (async actions are executed in the invoker thread until the first suspension point).
-     * While the C# approach seems to be more efficient, it ends up with recommendations like
-     * "use `yield` if you need to ....". This is error-prone. The JS-style approach is more consistent
-     * and does not require programmers to think about whether they need to yield or not.
+     * A dispatcher can override this method to provide a performance optimization and avoid paying a cost of an unnecessary dispatch.
+     * E.g. [MainCoroutineDispatcher.immediate] checks whether we are already in the required UI thread in this method and avoids
+     * an additional dispatch when it is not required. But this method should not return `false` by default, because
+     * it may expose unexpected behaviour (e.g. with interleaved event-loops) and unexpected order of events.
      *
-     * However, coroutine builders like [launch][CoroutineScope.launch] and [async][CoroutineScope.async] accept an optional [CoroutineStart]
-     * parameter that allows one to optionally choose the C#-style [CoroutineStart.UNDISPATCHED] behavior
-     * whenever it is needed for efficiency.
+     * Coroutine builders like [launch][CoroutineScope.launch] and [async][CoroutineScope.async] accept an optional [CoroutineStart]
+     * parameter that allows one to optionally choose the [undispatched][CoroutineStart.UNDISPATCHED] behavior to start coroutine immediately,
+     * but to be resumed only in the provided dispatcher.
      *
      * This method should generally be exception-safe. An exception thrown from this method
      * may leave the coroutines that use this dispatcher in the inconsistent and hard to debug state.
      *
-     * **Note: This is an experimental api.** Execution semantics of coroutines may change in the future when this function returns `false`.
+     * **This is an experimental api.** Execution semantics of coroutines may change in the future when this function returns `false`.
```

Comment:
Maybe it's time to drop experimental and leave it the way it is?

Reply:
On the one hand, we still change event order in event loops. On the other hand, `Unconfined` and `immediate` are stable and documentation states "order of events is an implementation detail that can be changed" . Let's drop experimental status